### ISSUE TITLE
[DevEx RFC] Improve the transaction builder pattern

### ISSUE THREAD

**Issue by davidiw** [aptos-labs/aptos-core#9335]
If you search around the various codebases, you'll see different approaches to transaction building.

Before we go there, let's talk about transactions:
* You have a RawTransaction that contains pretty boring stuff, so this is a straightforward builder pattern that largely exists within Rust
* You have your authenticator, which is a bit weird because they're quite specialized and you have to study them to figure out what you want to do -- single signer, multisigner, mutliagent, fee payer
* Different authenticators demand different types of raw transaction types
* No convenient way to ensure distinct parties are constructing the same transaction type

Arguably all of them are pretty bad:
* Each library let's users manually create transactions and sign them but extracting signatures is a bit awkward and so can be assembling authenticators from distinct processes
* Rust let's you make a transaction if you have all private keys locally

Potential solution:
* Generalize the core transaction builder and output a signable transaction -- not a real type
* Pass it around and receive back some sort of signed attestation (hash of txn, enum of signer type / index or address, and the signature)
* Generate signed transaction

This means we may need two more universal types -- bcs or json or something else... not great, but it would improve UX for complex, shared transactions.

I need to add links to this description.

If we do this in Rust first, we should really unify this into the types crate and eliminate all the duplicate transaction generation logic throughout the code.... I think both Python and TypeScript don't really have a good model for txn building at this point in time.

**Comment by xbtmatt** [comment]
I wanted to understand the transaction builder patterns myself last night, so I wrote this up to make sense of it.

### What a signed transaction consists of

There are a few different confusing terms here because they all sound similar but refer to different steps in the process of building a transaction.

A signed transaction is BCS serialized data where the data is comprised of a signed message and a raw transaction.
- A [raw transaction](https://github.com/aptos-labs/aptos-core/blob/main/ecosystem/typescript/sdk/src/aptos_types/transaction.ts) is all the details about what a transaction will *do* and logistical details about how to submit the transaction. It includes things like the sender's address, the sequence number, and the payload (the invoked entry function).
- A signed message is the encrypted hash formed from a private key signing a raw transaction. It represents a single account's approval to invoke an entry function that includes them as a `&signer`.
- A public key is the corresponding public key for the private key used to sign a transaction. It's used to decrypt a signed message into an unsigned message.

When a signed transaction is submitted, it must be validated by a validator node to ensure that it's a valid signed transaction. The basic process for this is ensuring that the signed message is well-formed; that is, that the raw transaction data matches the data that's output from decrypting the signed message with the public key.

An *AccountAuthenticator* is specifically the combination of the public key + a signed message. This represents each individual sender account's approval of a transaction.

There are multiple types of AccountAuthenticators, some of which include multiple accounts, which is what leads to the main differences in all of the builder patterns:

### Ed25519
- This is a basic, single signature account authenticator. It's just a public key and the signed message.
- One sender, one authenticator. In terms of its on-chain representation, it appears as one account aka a `&signer` primitive

### [MultiEd25519](https://aptos.dev/concepts/accounts/#multi-signer-authentication)
- a single sender and account verified with multiple (K of N) signatures. The initial authentication key aka its public address is constructed from hashing all N public keys with the required K number of signatures and the value 0x01 (the multied25519 domain separator). More succinctly: `auth_key = sha3-256(p_1 | . . . | p_n | K | 0x01)`
- In terms of an invoked entry function, there is only one `&signer` for a MultiEd25519 account.

### MultiAgent
- This authenticator scheme is complex, because it's represented as multiple layers of AccountAuthenticators. Technically there is only one sender in the raw transaction, but the parent authenticator scheme indicates that the transaction includes multiple signing accounts. This means within it is multiple individual authenticator schemes.
- This means there are multiple `&signer` primitives in the invoked function, for example: `0x3::token::direct_transfer_script(sender: &signer, receiver: &signer, ...)` where both the sender and the receiver are signers. The first signer is the primary address, aka the specified sender in the raw transaction.
- The secondary signer accounts/addresses are all authenticated with individual authenticators, potentially Ed25519 or even MultiEd25519 authenticators.
- To reiterate, all of the individual account authenticators are packed into the overarching MultiAgentAuthenticator like you can see in the typescript [directTransferToken example](https://github.com/aptos-labs/aptos-core/blob/main/ecosystem/typescript/sdk/src/plugins/token_client.ts)

### FeePayer
- This authenticator scheme is very similar to MultiAgent, but has a specific designated fee payer, separate from the secondary accounts. The main difference is that the fee payer is a unique type of secondary sender that does not show up as a `&signer` in the invoked entry function, it merely pays for the gas used to validate and execute the transaction.

TL;DR The parent authenticator type for a signed transaction indicates how a validator node should interpret the incoming BCS serialized transaction data. The individual authenticators within the overarching authenticator include the necessary data to verify that the message has been signed by the sender, any secondary accounts, and (now also) a potential fee payer.
